### PR TITLE
CB-15446: Added logic to SDX resize detach flow so that failures are recovered from.

### PR DESCRIPTION
**Pull request by sxxgrc**
Jira: https://jira.cloudera.com/browse/CB-15446

This was merged in before but was reverted due to an e2e issue. Previous PR: https://github.com/hortonworks/cloudbreak/pull/11965

### CODE REVIEW

**Review by horadla23 — APPROVED** [review]

**Review by tiborpopovics — APPROVED** [review]

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxDetachFailedEvent.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxDetachFailedEvent.java:18]: this method shouldn't be overwritten at all. `SdxEvent` has a default selector implementation which is just fine

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java:6]: the pain is coming from the fact that this class uses 
```SdxDetachEvent(Class eventClass) {
        this.event = eventClass.getSimpleName();
        this.event = eventClass.getSimpleName();
    }
    }
```
Instead of `EventSelectorUtil` which is the generic selector implementation for events...

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java:15]: private missing

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java:19]: public please and for the next method too

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:200]: so its not `detached`, it's should be called `reattached` at this point, or do I misunderstand the code here?

**Review by lacikaaa — CHANGES_REQUESTED** [review]

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java
Comment (lacikaaa) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java:6]: ah messed up the copy, but it's from the constructor code

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java:15]: This pattern is used by the test for this class which is why I didn't make this private.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java:19]: Sure I can do this.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:200]: No you're right, it goes from detached to reattached I just didn't want to create another object, but I can see how that can help with the clarity.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxDetachFailedEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxDetachFailedEvent.java:18]: I'll look into this, I was just sticking to the logic already present. Thank you for letting me know about this.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java:6]: Sorry could you expand on this a little? I am seeing inconsistent use of different logic throughout different events. The Start and Stop events use the same as what we are using here, and a lot of other events just use basic strings within the enum objects. 
Also that constructor is essentially not used anywhere so I am not sure how important it is to get it using something else.

I am open to suggestions though. This is my first flow change so I am new to a lot of this.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxDetachFailedEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxDetachFailedEvent.java:18]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java]: I modified it a little, feel free to take a look.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java
Comment (lacikaaa) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java:6]: sure, let me elaborate a bit

the problem with this approach this case have:
- it has a selector as a string in the event itself (`SdxDetachFailedEvent`)
- `SdxDetachEvent` calculates the selector based on a unique logic based on the class
If you rename the class but you forget about the string based selector, then this would only be recognized during runtime, as there is no IDE or compile time support.

Suggested way to handle selectors and event:
- for class based selectors use the `EventSelectorUtil`, and use it everywhere to ensure unity. So if you rename the class using the IDE it would rename it everywhere or it would cause a compile time error
- if you need string based selector, like you are reusing `SdxEvent` as you don't need anything else, then in this case in the `SdxDetachEvent` enum you would define an event with the string value, and when you create the event and set the selector you should use `SdxDetachEvent#getEvent`. This would ensure if the string changes for any reason it would still work

So the purpose is to use the same and proofed methods, and depend on the IDE and compiler as much as possible to ensure no errors are introduced if a change happens. I hope this clarifies, if not please reach out to me.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java:15]: Thank you for the information, I agree with you that basing the accessibility of a field or method off just testing does not fit the reason for seeking privacy in the first place.
I have chosen a solution that is kind of lamer than the ones you offered: I am choosing to just have another copy of the pattern in the test class.

The reason I am doing this is because my approach to testing is not just to ensure things work but also to ensure that in the future, if someone makes changes to the logic, the methods still function in the expected fashion.
So testing `generateNameFromDetached` by using a mock or by calling `generateOriginalNameFromDetached` does not really work for me since that doesn't really verify that the resulting name contains the original name plus an expected string.

Hopefully this makes sense and meets what you think is ideal.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachEvent.java:7]: Made changes to the `SdxDetachEvent` and `SdxDetachFailedEvent` classes. Thank you for the thorough information it will help me out a lot in the future.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:200]: I still think you should introduce a new variable here instead of reusing the old, as at this point the cluster is reattached. but this is not a major thing

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:319]: could you add a log here? I have seen really rarely that failure handling actions throw an error

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:57]: external calls shouldn't be part of a transaction as it could cause long db locks

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:120]: external calls shouldn't be part of a transaction

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:19]: nit: should be `@service` if the class is called service

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:102]: this should be private if I see it well

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java:26]: please replace generic exception with a CB specific one

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java:25]: nit: it would help readability if you invert the if, and the return would be in the `if` branch and the exception throwing logic in the `else` part (so you would have and else branch)

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java:50]: you should use `SdxService#getById` instead

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java:32]: you should use `SdxService` instead

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java:24]: should be service

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:27]: you should use `SdxService` instead

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:200]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:249]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:178]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:120]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:319]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:71]: Done. Let me know what you think about the naming for the exception.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:57]: Fixed.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:19]: Done but I would appreciate if you quickly explained the difference between a component and a service. I just did this because I saw similar files use component.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:102]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:27]: I am open to doing this but based on one of my recent changes addressing your comments, I may also want to **delete** a cluster from the repository. This is not exposed by the `SdxService`.
What do you think I should do with this in mind?

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:71]: Oh also I didn't know exactly what you meant by this so I ended up just creating a custom exception. 
If there is a specific CB exception I should be using instead please let me know.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java:26]: Done using the custom exception I created based on your comment above.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachNameGenerator.java:25]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java:50]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java:32]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java:24]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:27]: you should expose it using the service

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java]: no major difference, but if you call your class service then it makes sense to use a matching annotation. It's just a classification thing, that's why it's just a small thing, which wouldn't block merging a pr. I wouldn't have asked it to be changed if there wasn't any other issue
https://www.baeldung.com/spring-component-repository-service

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:126]: ok, thanks. let's wait for @kkalvagadda1 approval then

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachException.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachException.java:3]: nit: this could extend `CloudbreakServiceException`

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java]: I mentioned in a different comment, but you can extend or use `CloudbreakServiceException`

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (kkalvagadda1) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:126]: The majority of the places in the code, transaction service is used to wrap calls to repositories so that changes to the repository can be reverted when there are exceptions to achieve automicity.

Comment (kkalvagadda1) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:126]: I did not realize that assignResourceOwnerRoleIfEntitled makes a call to UMS service.
I'm good with this change. Let's access to update the owner assignment when resize is done. we could maintain the older owner assignment.

let's consider another PR for this.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachException.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachException.java]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:24]: this is not used anymore. please remove from the unit test too

**Review by lacikaaa — APPROVED** [review]

**Comment by horadla23** [comment]
before merge, pls check resize related test failure: https://jira.cloudera.com/browse/CB-16208

**Comment by sxxgrc** [comment]
Did a general rebase so this isn't too out of date and to aid in development of dependent changes.

**Comment by sxxgrc** [comment]
retest this please

**Pull request merged**
